### PR TITLE
CUDMABufferObject: add new class to abstract memfd/udmabuf buffers

### PR DESCRIPTION
**Pull request by lrusak**
This add using memfd/udmabuf's to the recent `BufferObjectFactory` that was in #17403 

This basically allows mapping userspace memory into the dma subsystem. This may or may not work on some systems and the memory may not be contiguous. Early testing also shows that this memory type is cached which is useful in the future for video buffers. We'll have to see how it goes and in the future may have to add criteria to the `BufferObjectFactory` to select specific `CBufferObjects` for specific tasks.

This will only be built and registered if your kernel is new enough to support `linux/udmabuf.h` which is version 4.20 (blaze it) and later.

To use this you need to be able to read/write to `/dev/udmabuf`. systemd tags this node via udev to the `kvm` group so your user needs to be a part of that or you need to use the following udev rule
```
KERNEL=="udmabuf", MODE="0666"
```
The `CUDMABufferObject` class will check permissions and only register if the permissions are correct.

This also will allow using wayland with dmabufs in the future. That will be coming later when we figure out some other stuff.

### CODE REVIEW

File: cmake/scripts/linux/ArchSetup.cmake [hunk 1]
```diff
@@ -88,6 +88,12 @@ check_include_files("linux/memfd.h" HAVE_LINUX_MEMFD)
 if(HAVE_LINUX_MEMFD)
   list(APPEND ARCH_DEFINES "-DHAVE_LINUX_MEMFD=1")
 endif()
+
+check_include_files("linux/udmabuf.h" HAVE_LINUX_UDMABUF)
```
Comment (garbear) [inline cmake/scripts/linux/ArchSetup.cmake]: Can you add a warning if this is missing so we can verify in compile logs after the fact?

**Comment by garbear** [comment]
That's pretty cool how you can add an implementation and push it to the stack, and all clients behind the buffer API get the benefit.

Out of curiosity, what are the performance implications of UDMA, GBM and dumb buffers? I would guess from the FILO that UDMA > GBM > dumb buffers

File: cmake/scripts/linux/ArchSetup.cmake (hunk 1 above)
Comment (lrusak) [reply cmake/scripts/linux/ArchSetup.cmake]: added

**Comment by lrusak** [comment]
> That's pretty cool how you can add an implementation and push it to the registration stack, and all clients behind the buffer API get the benefit.

Thanks! That's the idea :smile_cat: 

> Out of curiosity, what are the performance implications of UDMA, GBM and dumb buffers? I would guess from the FILO that UDMA > GBM > dumb buffers

Somewhat yes. Unfortunately documentation is a bit sparse for a bunch of it so we basically can categorize them as follows:

|   | CPU Cached | Supports DMA | Requires GBM | Contiguous |
|---|------------|--------------|--------------|--------------|
| UDMA |      :grey_question:     |       :white_check_mark:       |      :x:         | :grey_question: |
| GBM |       :x:      |       :white_check_mark:       |       :white_check_mark:       | :white_check_mark: |
| Dumb |      :x:       |       :white_check_mark:       |       :x:        | :white_check_mark: |

Now I'm not 100% sure if udma is cpu cached or not (it seems so with testing). Also udma uses memfd allocated memory which may not be contiguous (not 100% sure on that either). Typically systems that don't have an iommu cannot import non-contiguous memory into their GPU.

The only documentation I can find is [this](https://patchwork.kernel.org/patch/11108955/):
```
udmabuf is a device driver which allows userspace to create dmabufs.
The memory used for these dmabufs must be backed by memfd.
```

**Pull request merged**